kernel/devices/LApic.cpp

```diff
@@ -82,13 +101,41 @@ namespace Kernel::Devices
         //TODO: x2APIC support
         
         baseAddress = (CPU::ReadMsr(MSR_APIC_BASE) & ~(0xFFF)) + vmaHighAddr;
+        bool isX2ApicFeaturePresent = CPU::FeatureSupported(CpuFeature::X2APIC);
+        Logf("X2Apic feature status: %b", LogSeverity::Info, isX2ApicFeaturePresent);
         Memory::PageTableManager::Local()->MapMemory(baseAddress, baseAddress - vmaHighAddr, Memory::MemoryMapFlag::AllowWrites);
+        
         apicId = ReadReg(LocalApicRegister::Id) >> 24;
         timerTicksPerMs = 0;
 
+        uint64_t apic_msr_base = CPU::ReadMsr(MSR_APIC_BASE);
         if ((CPU::ReadMsr(MSR_APIC_BASE) & (1 << 11)) == 0)
             Log("IA32_APIC_BASE_MSR bit 11 (global enable) is cleared. Cannot initialize local apic.", LogSeverity::Fatal);
 
+        if(isX2ApicFeaturePresent) 
+        {
+            if (((apic_msr_base >> 10) & 0b11) != 0b11) 
+            {
+                Log("X2apic is not enabled", LogSeverity::Info);
+                x2ModeEnabled = isX2ApicFeaturePresent;
+                //To Enable x2apic we need to set the 10th bit in the MSR register and write it back.
+                Logf("Apic ms base before: %x\n", LogSeverity::Info, apic_msr_base);
+                apic_msr_base = apic_msr_base | (1 << 10);
+                Logf("Apic ms base after: %x\n", LogSeverity::Info, apic_msr_base);
+                CPU::WriteMsr(MSR_APIC_BASE, apic_msr_base);
+                uint64_t x2apic_id = CPU::ReadMsr(0x802);
+                Logf("Apic id: %x\n", LogSeverity::Info, x2apic_id);
+                Logf("Test conversion Spurious: %x\n", LogSeverity::Info, (((uint64_t) LocalApicRegister::SpuriousInterruptVector)>> 4) + 0x800);
+                Logf("Test macro: %x\n", LogSeverity::Info, ExtractX2Offset(LocalApicRegister::SpuriousInterruptVector));
+                Logf("Test conversion Timer: %x\n", LogSeverity::Info, (((uint64_t) LocalApicRegister::TimerLVT)>> 4) + 0x800);
+                Logf("Test macro: %x\n", LogSeverity::Info, ExtractX2Offset(LocalApicRegister::TimerLVT));
+                Logf("Test macro EOI: %x\n", LogSeverity::Info, ExtractX2Offset(LocalApicRegister::EOI));
+
+            }  
+
+            x2ModeEnabled = true;
+        }
```

Comment:
I understand the use of logging for debugging, but before I merge the PR perhaps we should trim it a little :)
Maybe what would be good is that only the bsp outputs a single log line depending on which mode it's configured in (you can check for this with `IsBsp()`), and at LogSeverity::Verbose, not Info (so we can filter it as non-critical info later on)
Maybe it would output something like:
if x2 is available, and already enabled by bootloader = "System is using X2 APIC, initialized by bootloader"
if available, and we enabled it = "System successfully initialized X2 APIC."
if not available = "Local APICs running XAPIC mode."